src/stan/analyze/mcmc/compute_effective_sample_size.hpp

```diff
@@ -1,98 +1,184 @@
+    rho_hat_s(1) = rho_hat_odd;
+    // Geyer's initial positive sequence
+    size_t s = 1;
+    for (; s < (num_draws - 4) && (rho_hat_even + rho_hat_odd) > 0;
+         s += 2) {
       for (int chain = 0; chain < num_chains; ++chain)
-        acov_s(chain) = acov(chain)(1);
-      double rho_hat_even = 1;
-      double rho_hat_odd = 1 - (mean_var - acov_s.mean()) / var_plus;
-      rho_hat_s(1) = rho_hat_odd;
-      // Geyer's initial positive sequence
-      int max_s = 1;
-      for (size_t s = 1;
-           (s < (num_draws - 2) && (rho_hat_even + rho_hat_odd) >= 0);
-           s += 2) {
-        for (int chain = 0; chain < num_chains; ++chain)
-          acov_s(chain) = acov(chain)(s + 1);
-        rho_hat_even = 1 - (mean_var - acov_s.mean()) / var_plus;
-        for (int chain = 0; chain < num_chains; ++chain)
-          acov_s(chain) = acov(chain)(s + 2);
-        rho_hat_odd = 1 - (mean_var - acov_s.mean()) / var_plus;
-        if ((rho_hat_even + rho_hat_odd) >= 0) {
-          rho_hat_s(s + 1) = rho_hat_even;
-          rho_hat_s(s + 2) = rho_hat_odd;
-        }
-        max_s = s + 2;
-      }
-      // Geyer's initial monotone sequence
-      for (int s = 3; s <= max_s - 2; s += 2) {
-        if (rho_hat_s(s + 1) + rho_hat_s(s + 2) >
-            rho_hat_s(s - 1) + rho_hat_s(s)) {
-          rho_hat_s(s + 1) = (rho_hat_s(s - 1) + rho_hat_s(s)) / 2;
-          rho_hat_s(s + 2) = rho_hat_s(s + 1);
-        }
+        acov_s(chain) = acov(chain)(s + 1);
+      rho_hat_even = 1 - (mean_var - acov_s.mean()) / var_plus;
+      for (int chain = 0; chain < num_chains; ++chain)
+        acov_s(chain) = acov(chain)(s + 2);
+      rho_hat_odd = 1 - (mean_var - acov_s.mean()) / var_plus;
+      if ((rho_hat_even + rho_hat_odd) >= 0) {
+        rho_hat_s(s + 1) = rho_hat_even;
+        rho_hat_s(s + 2) = rho_hat_odd;
       }
+    }
 
-      return num_chains * num_draws / (1 + 2 * rho_hat_s.sum());
+    int max_s = s;
+    // this is used in the improved estimate
+    if (rho_hat_even > 0)
+      rho_hat_s(max_s + 1) = rho_hat_even;
+
+    // Geyer's initial monotone sequence
+    for (int s = 1; s <= max_s - 3; s += 2) {
+      if (rho_hat_s(s + 1) + rho_hat_s(s + 2) >
+          rho_hat_s(s - 1) + rho_hat_s(s)) {
+        rho_hat_s(s + 1) = (rho_hat_s(s - 1) + rho_hat_s(s)) / 2;
+        rho_hat_s(s + 2) = rho_hat_s(s + 1);
+      }
     }
 
-    inline
-    double compute_effective_sample_size(std::vector<const double*> draws,
-                                         size_t size) {
-      int num_chains = draws.size();
-      std::vector<size_t> sizes(num_chains, size);
-      return compute_effective_sample_size(draws, sizes);
+    double ess = num_chains * num_draws;
+    // Geyer's truncated estimate
+    // Improved estimate reduces variance in antithetic case
+    double tau_hat = -1 + 2*rho_hat_s.head(max_s).sum() + rho_hat_s(max_s + 1);
+    // Safety check for negative values and
```

Comment:
I think that this threshold is worth noting in the function documentation itself.

Reply:
Would you give a look over my latest efforts to document this?